### PR TITLE
split DistributedUtil into new `distributed` package

### PR DESCRIPTION
**Pull request by ryan-williams**
* pileup flatmap functions → `PileupFlatMapUtils`
* window flatmap functions → `WindowFlatMapUtils`
* loci partitioning functions → `LociPartitioningUtils`
* other miscellaneous classes into their own files

---
This change is [Reviewable](https://reviewable.io/reviews/hammerlab/guacamole/427)

### CODE REVIEW

File: src/main/scala/org/hammerlab/guacamole/dist/WindowFlatMapUtils.scala [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+    * @param function function to flatMap: (task number, loci, iterators of regions that overlap a window around these
+    *                 loci (one region-iterator per sample)) -> T
+    * @tparam T type of value returned by function
+    * @return flatMap results, RDD[T]
+    */
+  private[dist] def windowTaskFlatMapMultipleRDDs[M <: HasReferenceRegion: ClassTag, T: ClassTag](
+    regionRDDs: PerSample[RDD[M]],
+    lociPartitions: LociMap[Long],
+    halfWindowSize: Long,
+    // TODO(ryan): factor this function type out (as a PartialFunction?)
```
Comment (arahuja) [inline src/main/scala/org/hammerlab/guacamole/dist/WindowFlatMapUtils.scala]: Worth removing this TODO - is it still a todo? Or make an issue for it instead?

**Comment by arahuja** [comment]
👍  Like this change!

File: src/main/scala/org/hammerlab/guacamole/dist/WindowFlatMapUtils.scala (hunk 1 above)
Comment (ryan-williams) [inline src/main/scala/org/hammerlab/guacamole/dist/WindowFlatMapUtils.scala]: good idea! done

**Comment by timodonnell** [comment]
Yeah, this looks good to me @ryan-williams . My only thought is that ‘dist’ package name may be somewhat unclear (my first thought is it has something to do with package management), but it is tough to come up with a clear name. Maybe `distributed_utils` or `distributed_primitives`? Then could also drop the ‘Utils’ from the object names perhaps.

**Comment by ryan-williams** [comment]
Good point, I'd support changing the package name to `utils` instead of `dist`. I prefer keeping the `Utils` suffixes on the `object`s though.

**Comment by ryan-williams** [comment]
Actually getting in to the code, I decided to package them more granularly:
* `LociPartitionUtils` → new `loci` package.
* `WindowFlatMapUtils` → `windowing` package.
* `PileupFlatMapUtils` → `pileup` package.

I also put `LociMap` and `LociSet` into the new `loci` package.

**Comment by timodonnell** [comment]
I see the logic here, but I liked the separation between Spark-independent functionality like `Pileup` and Spark-based distributed primitives like `pileupFlatmap`. @arahuja has mentioned even splitting these up into their own modules (#329), since there may be use cases for the basic functionality in non-Spark contexts. So my vote would be your original setup in this PR with a `dist` package, except call that package something a little clearer like `distributed` (I'm not a fan of calling it `utils`, since that doesn't make it clear this is the sub-package where distributed operations are implemented using Spark).

**Comment by ryan-williams** [comment]
OK, moved everything into a new `distributed` package

**Comment by ryan-williams** [comment]
Going to merge this in, lmk if there are further comments!

**Pull request merged**